### PR TITLE
Add support for coercions in table scan redirections

### PR DESCRIPTION
**Pull request by sopel39**
Depends on: https://github.com/prestosql/presto/pull/6030

### CODE REVIEW

File: presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java [hunk 1]
```diff
@@ -201,6 +211,25 @@ public void testRedirectAfterColumnPruningOnPushedDownPredicate()
...
+    @Test
+    public void testPredicateTypeMismatch()
+    {
+        try (LocalQueryRunner queryRunner = createLocalQueryRunner(
+                getMockApplyRedirectAfterPredicatePushdown(typeMismatchedRedirectionMappingBC, Optional.of(ImmutableSet.of(sourceColumnHandleB))),
+                Optional.of(this::mockApplyProjection),
+                Optional.of(getMockApplyFilter(ImmutableSet.of(sourceColumnHandleC))))) {
+            // After 'source_col_c = 1' is pushed into source table scan, it's possible for 'source_col_c' table scan assignment to be pruned
+            // Redirection results in Project('dest_col_b') -> Filter('dest_col_c = 1') -> TableScan for such case
+            // but dest_col_a has mismatched type compared to source domain
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java]: What's `dest_col_a` in this example?

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java [hunk 2]
```diff
@@ -16,6 +16,8 @@
 import com.google.common.collect.ImmutableBiMap;
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java:16]: Commit message should provide some explanation on what this change is about

File: presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java [hunk 3]
```diff
@@ -26,12 +26,27 @@
     private final Map<ColumnHandle, String> destinationColumns;
     // filter that needs to be applied on top of table scan
     private final TupleDomain<String> filter;
+    // should coercions be used when source and redirected column types don't match
+    private boolean allowCoercions;
```
Comment (martint) [inline presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java:30]: When and why would this ever be false? This doesn't seem necessary. If the connector doesn't want to allow coercions, it should just not perform the redirection.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java [hunk 4]
```diff
@@ -215,12 +223,12 @@ public void testRedirectAfterColumnPruningOnPushedDownPredicate()
...
     {
         try (LocalQueryRunner queryRunner = createLocalQueryRunner(
-                getMockApplyRedirectAfterPredicatePushdown(typeMismatchedRedirectionMappingBC, Optional.of(ImmutableSet.of(sourceColumnHandleB))),
+                getMockApplyRedirectAfterPredicatePushdown(typeMismatchedRedirectionMappingBC, Optional.of(ImmutableSet.of(sourceColumnHandleB)), false),
                 Optional.of(this::mockApplyProjection),
                 Optional.of(getMockApplyFilter(ImmutableSet.of(sourceColumnHandleC))))) {
             // After 'source_col_c = 1' is pushed into source table scan, it's possible for 'source_col_c' table scan assignment to be pruned
             // Redirection results in Project('dest_col_b') -> Filter('dest_col_c = 1') -> TableScan for such case
-            // but dest_col_a has mismatched type compared to source domain
+            // but dest_col_c has mismatched type compared to source domain
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java:231]: This belongs in the previous commit

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java [hunk 5]
```diff
@@ -88,49 +94,56 @@ public Result apply(TableScanNode scanNode, Captures captures, Context context)
...
-                    }
-
-                    return destinationColumnHandle;
-                }));
+        ImmutableMap.Builder<Symbol, Cast> casts = ImmutableMap.builder();
+        ImmutableMap.Builder<Symbol, ColumnHandle> newAssignmentsBuilder = ImmutableMap.builder();
+        for (Map.Entry<Symbol, ColumnHandle> assignment : scanNode.getAssignments().entrySet()) {
+            String destinationColumn = columnMapping.get(assignment.getValue());
+            if (destinationColumn == null) {
+                throw new PrestoException(COLUMN_NOT_FOUND, format("Did not find mapping for source column %s in table scan redirection", assignment.getValue()));
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java:102]: TrinoException

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java [hunk 6]
```diff
@@ -144,70 +157,118 @@ public Result apply(TableScanNode scanNode, Captures captures, Context context)
...
+                    "Cast not possible from redirected column %s.%s with type %s to source column %s.%s with type: %s",
+                    destinationTable,
+                    destinationColumn,
+                    destinationType,
+                    sourceTable,
+                    sourceColumnHandle,
+                    sourceType));
+        }
+
+        return new Cast(destinationSymbol.toSymbolReference(), toSqlType(sourceType));
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java:272]: This is missing the typeOnlyCoercion flag.

**Review by sopel39 — COMMENTED** [review]
rebased. PTAL

File: presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java (hunk 1 above)
Comment (sopel39) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java]: `dest_col_a` is integer

File: presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java (hunk 3 above)
Comment (martint) [reply presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java]: The can do this on its own. It needs to somehow be aware of the target table anyway, so it can do this check by itself.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java (hunk 6 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java]: I know. But currently it's saying every coercions is "not type-only coercion". It needs to pass the true/false flag depending on whether that's the case.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java (hunk 1 above)
Comment (martint) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java]: Sorry, I meant what role does it play since it seems unrelated to the example given. But then I noticed you changed this to say `dest_col_c` in another commit.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java (hunk 6 above)
Comment (findepi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java]: @martint shouldn't this not be a concern of this rule?
ie shouldn't we have a canonicalizer which can determine a certain cast is type-only?
it's a function of types being converted from/to, not a function of particular creating site

Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java]: Well, the notion of type-only casts in the planner has always been somewhat of a misfeature. It's not clear it even does anything (a type-only cast will be just a passthrough method in the implementation, so the JIT should be able to inline and remove them).

But we don't have any such thing today, and we rely on the callsites to construct the cast properly. In a way, you can think of an IR cast as having `(from, to, type-only)` elements (vs the AST cast, which only has (`from`, `to`))

TL;DR, we should consider getting rid of type-only casts, but in the meantime, we should construct them properly if at all possible.

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java]: > TL;DR, we should consider getting rid of type-only casts,

can you please file an issue for that?

File: presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java (hunk 1 above)
Comment (sopel39) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java]: `source_col_c` gets redirected to `dest_col_a`, but `dest_col_a` has different type (int) than `source_col_c` (varchar)

File: presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java (hunk 3 above)
Comment (sopel39) [reply presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java]: >  It needs to somehow be aware of the target table anyway

It might not be aware of target table types since it can be in different catalog and could be created externally.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java (hunk 6 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java]: https://github.com/trinodb/trino/issues/8111

File: presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java (hunk 3 above)
Comment (martint) [reply presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java]: Then how is it going to know to redirect to those tables, or that they are the right table to redirect to? Table redirections don't seem very useful without that level of understanding, so it seems reasonable to me to expect that from connectors.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java (hunk 6 above)
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java]: added call to `isTypeOnlyCoercion`

File: presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java (hunk 3 above)
Comment (sopel39) [reply presto-spi/src/main/java/io/prestosql/spi/connector/TableScanRedirectApplicationResult.java]: removed toggle

**Review by martint — APPROVED** [review]
